### PR TITLE
Add TypeScript definition

### PR DESCRIPTION
**Pull request by Fazendaaa**
Hey, @sindresorhus.

Due to Hacktoberfest 2018, I'm contributing to some packages.

TypeScript typings help a lot of us who use it due to it's integration to TSLint and some more strict checkings, needed to be performed in some cases in the used packages as well; that's why I've decided to share the ones that I've made, making it more accessible to all programmers in this environment :)

Besides adding this, I would love to add a Husky integration to perform linter checkings, tests and documentation to git hooks. While doing this, if is it of your interest I can rewrite the JS parts in TS also, adding tests and everything.

### CODE REVIEW

**Comment by sindresorhus** [comment]
Hey. Awesome to see participation for Hacktoberfest. ✨

Can you follow [this styleguide](https://github.com/sindresorhus/typescript-definition-style-guide)?

**Comment by sindresorhus** [comment]
> Besides adding this, I would love to add a Husky integration to perform linter checkings, tests and documentation to git hooks. While doing this, if is it of your interest I can rewrite the JS parts in TS also, adding tests and everything.

I appreciate the offer, but I'm not interested in that.

**Comment by Fazendaaa** [comment]
> Hey. Awesome to see participation for Hacktoberfest. sparkles
> 
> Can you follow [this styleguide](https://github.com/sindresorhus/typescript-definition-style-guide)?

No problems :)

**Comment by Fazendaaa** [comment]
@sindresorhus what awesome guideline for contributions, I did not know these standards of yours and the first time I see something like that. I learned a lot.

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+export interface ShellEnv {
+    readonly [x: string]: string;
+    readonly SHELL: string;
+    readonly TERM_PROGRAM: string;
```
Comment (sindresorhus) [inline index.d.ts]: These are no guaranteed to exist.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+export interface ShellEnv {
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
export interface EnvironmentVariables {
```

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    readonly [x: string]: string;
```
Comment (sindresorhus) [inline index.d.ts]:
```suggestion
    readonly [key: string]: string;
```

File: package.json [hunk 4]
```diff
@@ -39,6 +40,12 @@
 	},
 	"devDependencies": {
 		"ava": "*",
+		"tsd-check": "^0.2.1",
 		"xo": "*"
+	},
+	"xo": {
+		"ignores": [
+			"*.ts"
+		]
```
Comment (sindresorhus) [inline package.json]: This is not needed.

File: index.test-d.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import {expectType} from 'tsd-check';
+import shellEnv, { sync, ShellEnv } from '.';
+
+(async () => {
```
Comment (sindresorhus) [inline index.test-d.ts]: > tsd-check supports top-level await. - https://github.com/sindresorhus/typescript-definition-style-guide

File: index.test-d.ts [hunk 6]
```diff
@@ -0,0 +1,7 @@
+import {expectType} from 'tsd-check';
+import shellEnv, { sync, ShellEnv } from '.';
```
Comment (sindresorhus) [inline index.test-d.ts]:
```suggestion
import shellEnv, {sync, ShellEnv} from '.';
```

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+ */
+export function sync(shell?: string): ShellEnv;
+
+/**
+ * Get the environment variables defined in your dotfiles.
+ * 
+ * @param shell To read the environment variables from.
+ * @returns The environment variables.
+ */
+export function shellEnv(shell?: string): Promise<ShellEnv>;
```
Comment (sindresorhus) [inline index.d.ts]: This should be a default export only, not a named export too.

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+ *
+ * @param shell To read the environment variables from.
+ * @returns The environment variables.
+ */
+export function sync(shell?: string): ShellEnv;
+
+/**
+ * Get the environment variables defined in your dotfiles.
+ * 
+ * @param shell To read the environment variables from.
```
Comment (sindresorhus) [inline index.d.ts]: > There's a dash after the parameter name. - https://github.com/sindresorhus/typescript-definition-style-guide

Comment (sindresorhus) [reply index.d.ts]:
```suggestion
 * @param shell - The shell to read the environment variables from.
```

**Comment by sindresorhus** [comment]
> For packages with a default export, use export default foo;, not export = foo;. You will have to add module.exports.default = foo; to the index.js file.

> The pull request should have the title Add TypeScript definition. (Copy-paste it so you don't get it wrong)

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
+export interface ShellEnv {
+    readonly [x: string]: string;
+    readonly SHELL: string;
+    readonly TERM_PROGRAM: string;
+}
+
+/**
+ * Get the environment variables defined in your dotfiles.
+ *
+ * @param shell To read the environment variables from.
```
Comment (sindresorhus) [inline index.d.ts]: Document the default value.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (Fazendaaa) [reply package.json]: Okay

File: index.d.ts (hunk 1 above)
Comment (Fazendaaa) [reply index.d.ts]: Thanks

**Review by sindresorhus — APPROVED** [review]

**Comment by sindresorhus** [comment]
Nice work! :)

**Pull request merged**